### ISSUE TITLE
Help needed training mnist lenet5 inspired topology

### ISSUE THREAD

**Issue by blackccpie** [tiny-dnn/tiny-dnn#718]
Hi, I've started working on an mnist training app, based on tiny-dnn's mnist example code, an trying to reproduce this network topology : [keras-cnn-inspired-by-lenet-5](https://www.kaggle.com/ftence/keras-cnn-inspired-by-lenet-5/code/code).
My app source code is [here](https://github.com/blackccpie/tiny-brain/tree/master/apps/mnist_autotrain), but I'm facing some issues trying to make training converge. Using the mnist dataset included in tiny-dnn, after 6 epochs the scores drops from 9832/10000 to 980/10000.
As you can see I use a custom adamax optimizer.

Could you help me understand this drop? Did I make something wrong in my code?

Thx for the help

**Comment by Randl** [comment]
have you tried reducing learning rate?

**Comment by blackccpie** [comment]
not yet, I started experimenting with the same adamax initial learning rate as keras, that is  [0.002](https://keras.io/optimizers/#adamax). I'm gonna have a few tries with lower ones.

**Comment by blackccpie** [comment]
Still no luck trying lower learning rates. My feeling is that it may come from exploding gradients, due to the cross entropy multiclass formulation `-t[i] / y[i]`, where `y[i]` may turn very small.
I saw people using simplified gradient formulation for the cross entropy with softmax use case, like in this [article](http://peterroelants.github.io/posts/neural_network_implementation_intermezzo02), that may perhaps solve some numerical instabilities?

**Comment by beru** [comment]
There's a similar problem with `cross_entropy::df` method in `tiny_dnn/lossfunctions/loss_function.h`.
https://github.com/tiny-dnn/tiny-dnn/issues/570

**Comment by blackccpie** [comment]
Thx for the tip, I tried the max with epsilon trick in the cross entropy multiclass derivative, but instead of a brutal drop I get a slow decrease. So think I'll stick with my current version for now, as an 98.3% score is enough for my needs.

**Comment by jnamdarirani** [comment]
I'm experiencing these exact same drops on cifar10 and mnist when I add some batch norm layers to the example training file. Here is the net containing the layers I added:
```
  nn << conv(32, 32, 5, 3, n_fmaps, padding::same, true, 1, 1,
             backend_type)                      // C1
     << batch_normalization_layer(32*32, n_fmaps) // BN1
     << pool(32, 32, n_fmaps, 2, backend_type)  // P2
     << relu()                                  // activation
     << conv(16, 16, 5, n_fmaps, n_fmaps, padding::same, true, 1, 1,
             backend_type)                      // C3
     << batch_normalization_layer(16*16, n_fmaps) // BN2
     << pool(16, 16, n_fmaps, 2, backend_type)  // P4
     << relu()                                  // activation
     << conv(8, 8, 5, n_fmaps, n_fmaps2, padding::same, true, 1, 1,
             backend_type)                                // C5
     << batch_normalization_layer(8*8, n_fmaps2) // BN2
     << pool(8, 8, n_fmaps2, 2, backend_type)             // P6
     << relu()                                            // activation
     << fc(4 * 4 * n_fmaps2, n_fc, true, backend_type)    // FC7
     << relu()                                            // activation
     << fc(n_fc, 10, true, backend_type) << softmax(10);  // FC10
```

The default example w/o the BN layers improves accuracy every epoch (I got to 60+% accuracy under 10 epochs with default parameters). With the BN layers I get a better accuracy at the end of the first epoch and then it decreases upon reaching random guess. Should I open a new issue about this?

**Comment by blackccpie** [comment]
what loss function are you using for training? If you're using cross entropy, then your problem might be related to mine.

**Comment by StripedBanana** [comment]
I didn't change it so it's the same as in the vanilla example.
